src/io/cached_itarbundle.rs

```diff
@@ -0,0 +1,716 @@
+// src/io/cached_itarbundle.rs -- I/O on files in an indexed tar file "bundle" cached locally
+// Copyright 2017-2019 the Tectonic Project
+// Licensed under the MIT License.
+
+use flate2::read::GzDecoder;
+use reqwest::{header::HeaderMap, Client, RedirectPolicy, Response, StatusCode};
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::io::{BufRead, BufReader, Read, Write};
+
+use crate::errors::{Error, ErrorKind, Result, ResultExt};
+use crate::status::StatusBackend;
+use crate::{ctry, tt_note, tt_warning};
+
+use fs2::FileExt;
+use std::fs::{self, File};
+use std::io::ErrorKind as IoErrorKind;
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use super::{try_open_file, Bundle, InputHandle, InputOrigin, IoProvider, OpenResult};
+use crate::digest::{self, Digest, DigestData};
+
+const MAX_HTTP_REDIRECTS_ALLOWED: usize = 10;
+const MAX_HTTP_ATTEMPTS: usize = 4;
+
+/// A simple way to read chunks out of a big seekable byte stream. You could
+/// implement this for io::File pretty trivially but that's not currently
+/// needed.
+pub struct HttpRangeReader {
+    url: String,
+    client: Client,
+}
+
+impl HttpRangeReader {
+    pub fn new(url: &str) -> HttpRangeReader {
+        HttpRangeReader {
+            url: url.to_owned(),
+            client: Client::new(),
+        }
+    }
+}
+
+impl HttpRangeReader {
+    fn read_range(&mut self, offset: u64, length: usize) -> Result<Response> {
+        let end_inclusive = offset + length as u64 - 1;
+
+        let mut headers = HeaderMap::new();
+        use headers::HeaderMapExt;
+        headers.typed_insert(headers::Range::bytes(offset..=end_inclusive).unwrap());
+
+        let res = self.client.get(&self.url).headers(headers).send()?;
+
+        if res.status() != StatusCode::PARTIAL_CONTENT {
+            return Err(Error::from(ErrorKind::UnexpectedHttpResponse(
+                self.url.clone(),
+                res.status(),
+            )))
+            .chain_err(|| format!("read range expected {}", StatusCode::PARTIAL_CONTENT));
+        }
+
+        Ok(res)
+    }
+}
+
+#[derive(Clone, Copy)]
+struct FileInfo {
+    offset: u64,
+    length: u64,
+}
+
+struct LocalCacheItem {
+    _length: u64,
```

Comment:
what is the significance of underscoring the field here?

Reply:
The field is never read in the code I believe. The manifest stores the length of the files but tectonic never uses the information. (This part is taken from `local_cache.rs`.)